### PR TITLE
mgr/orchestrator: Add service_action method

### PR DESCRIPTION
**Pull request by tserong**
This is to facilitate service start/stop/reload (for example,
after altering NFS Ganesha configuration, we need to trigger
a service reload).

Signed-off-by: Tim Serong <tserong@suse.com>

### CODE REVIEW

**Comment by tserong** [comment]
Just for the sake of defining the API endpoint. This could alternately be done with separate start/stop/reload functions, but I figured a single function to perform an operation of this type probably results in less code for everyone. @jmolmo @jtlayton @sebastian-philipp any opinions, concerns, etc.? If it looks OK, I'm happy to go do the matching piece in orchestrator_cli unless anyone's already working on that (and in deepsea too, of course).

**Comment by jmolmo** [comment]
LGTM! Thank you.

File: src/pybind/mgr/orchestrator.py [hunk 1]
```diff
@@ -183,6 +183,18 @@ def describe_service(self, service_type=None, service_id=None, node_name=None):
         """
         raise NotImplementedError()
 
+    def service_operation(self, operation, service_type, service_id=None):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: Is there a better or more precise word for "`operation`", something that excludes adding and removing services?

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -183,6 +183,18 @@ def describe_service(self, service_type=None, service_id=None, node_name=None):
...
+        """
+        Perform an operation (start/stop/reload) on a service.
+
+        If a specific service_id is passed in, perform the op on that service
+        instance, otherwise perform the op on all services of the specified
+        type.
+
+        :return: A WriteCompletion object
+        """
+        raise NotImplementedError()
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: can we add
```python
assert operation in ["start", "stop", "reload"]
```
?

File: src/pybind/mgr/orchestrator.py (hunk 1 above)
Comment (tserong) [reply src/pybind/mgr/orchestrator.py]: I don't know, but I'm totally open to suggestions ("operation" was the least worst thing I could think of :-/)

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (tserong) [reply src/pybind/mgr/orchestrator.py]: Yep, sure.

File: src/pybind/mgr/orchestrator.py (hunk 1 above)
Comment (tserong) [reply src/pybind/mgr/orchestrator.py]: @sebastian-philipp, what about "action" instead?  It's a bit narrower in scope.

**Comment by tserong** [comment]
I've reworked this slightly:

- Use the term "action" instead of "operation" ("operation" is used elsewhere in comments in a broader context, so best to have a different term).
- `service_action()` lets you operate either on a logical service or a specific instance, per the Orchestrator CLI spec google doc
- I've added matching commands to orchestrator_cli

Note: I've cherry-picked #25492 temporarily and will rebase once that PR is merged into master.

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 3]
```diff
@@ -247,8 +232,28 @@ def _service_rm(self, cmd):
         svc_type = cmd['svc_type']
         svc_id = cmd['svc_id']
 
-        completion = self._oremote("remove_stateless_service", svc_type, svc_id)
-        self._wait([completion])
+        completion = self.remove_stateless_service(svc_type, svc_id)
+        self._orchestrator_wait([completion])
+        return HandleCommandResult(rs="Success.")
+
+    def _service_action(self, cmd):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator_cli/module.py]: Could you add a simple test to `qa/tasks/mg/test_orchestrator_cli` that executes this code? Doesn't need to perform any action.

File: src/pybind/mgr/orchestrator.py [hunk 4]
```diff
@@ -183,6 +186,23 @@ def describe_service(self, service_type=None, service_id=None, node_name=None):
         """
         raise NotImplementedError()
 
+    def service_action(self, action, service_type, service_name=None, service_id=None):
```
Comment (rjfd) [inline src/pybind/mgr/orchestrator.py]: it's not clear to me the difference between `service_type` and `service_name`. Could you give a real example of how these two parameter are used with different values?

**Comment by rjfd** [comment]
@tserong I believe it would be worthwhile to add some docstrings to the `add_stateless_service`, `update_stateless_service`, and `remove_stateless_service` methods because it's not clear what is the semantics of those methods in the presence of a `service_action` method.

For instance, I was using the `add_stateless.../remove_stateless...` methods to start and stop the nfs-ganesha daemon here: https://github.com/rjfd/ceph/blob/wip-dashboard-nfs/src/pybind/mgr/ssh-orch/module.py#L455

**Comment by rjfd** [comment]
> @rjfd adding and removing stateless services is about deploying and "undeploying" them, i.e. if you add a service, you're deploying a new thing, and if you remove it, you're getting rid of it (think: uninstalling). That's not the same thing as starting and stopping a currently deployed service.

@tserong I fully agree with you. That's why I think we should add that description to those methods

**Comment by sebastian-philipp** [comment]
might make sense to move the docstrings into a separate PR. That would be a bit easier to merge.

**Comment by rjfd** [comment]
fine by me

**Comment by sebastian-philipp** [comment]
> fine by me

See #25669

File: src/pybind/mgr/orchestrator.py (hunk 4 above)
Comment (jtlayton) [reply src/pybind/mgr/orchestrator.py]: AIUI, type would be something like "nfs" or "mds". The bigger distinction is between "name" and "id" here, and I think that's more or less described in the delta below.

**Comment by tserong** [comment]
> Functionality looks good overall, but I think @rjfd has a good point that we may need to somehow make a clearer distinction between service_id and service_name (and likely, across the entire orchestrator).

Yeah...  This is explained in the orchestrator CLI google doc, but I don't know that the distinction has really made it into the in-tree docs or the orchestrator code itself yet.

`service_type` is just the type of service ("nfs", "mds", "rgw", whatever).

`service_id` is the ID of a specific instance of a running daemon of whatever type.

`service_name` is the name of a logical service, which may consist of many running daemons. This could be, for example, an RGW zone name, or a CephFS filesystem name.

**Comment by sebastian-philipp** [comment]
QA successful. Failure tracked in http://tracker.ceph.com/issues/37736

http://pulpito.ceph.com/swagner-2019-01-15_11:15:44-rados:mgr-wip-swagner-testing-distro-basic-smithi/

**Pull request merged**